### PR TITLE
NO2 emission and COVID Lockdown effects

### PR DESCRIPTION
**Pull request by Pratichhya**
In this notebook, we aim to explore the trend in air pollution, focusing on the NO2 product of the Sentinel 5P collection. We will, compare the situation before and after the onset of COVID-19 lockdown in the Delhi region of India.

### CODE REVIEW

File: openeo/Pre&PostCovid.ipynb [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "3bf6dcb3-dbdc-4ec9-9248-29977999f2bb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# import necessary packages\n",
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:29]:
```suggestion
```

that's not a very useful comment

File: openeo/Pre&PostCovid.ipynb [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    "import openeo\n",
+    "from openeo.processes import min, max, mean\n",
+    "import scipy\n",
+    "import numpy as np"
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:33]:
```
from openeo.processes import min, max, mean
import scipy
import numpy as np
```

none of these imports seems to used

File: openeo/Pre&PostCovid.ipynb [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    "                        [77.29, 28.56],\n",
+    "                        [77.29, 28.69],\n",
+    "                        [77.11, 28.69],\n",
+    "                    ]\n",
+    "                ],\n",
+    "                \"type\": \"Polygon\",\n",
+    "            },\n",
+    "        }\n",
+    "    ],\n",
+    "}"
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:89]: You use a feature collection here, but you can just use the inner polygon
that will save on a lot of non-interesting lines of code

File: openeo/Pre&PostCovid.ipynb [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    "# Now create a spatial aggregation to generate mean timeseries data\n",
+    "s5post = s5post.aggregate_spatial(reducer=\"mean\", geometries=aoi)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c670bad3-00a5-44f2-8104-99e78e16098e",
+   "metadata": {},
+   "source": [
+    "Finally we execute them as batch jobs and download the results as CSVs. They are further plotted as shown below to study the data."
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:159]: you say CSV here, but use file extension ".nc", suggesting netCDF

File: openeo/Pre&PostCovid.ipynb [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "e4f41bef-5136-4e51-a1b3-34782c728df8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import xarray as xr\n",
+    "import numpy as np\n",
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:244]: numpy is not used I think

File: openeo/Pre&PostCovid.ipynb [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "fig, ax1 = plt.subplots(dpi=100)\n",
+    "ax1.plot(predata.t, predata.NO2.to_numpy().flatten(), color=\"g\")\n",
+    "ax1.set_xlabel(\"During COVID\")\n",
+    "ax2 = ax1.twiny()\n",
+    "ax2.plot(postdata.t, postdata.NO2.to_numpy().flatten(), color=\"r\")\n",
+    "ax2.set_xlabel(\"Post COVID\")"
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:312]: add a legend to the plot clearly indicating that the first (green) series is "During covid" and the second (red) is "Post covid"

Comment (soxofaan) [reply openeo/Pre&PostCovid.ipynb:312]: I would also add a y label "NO_2" or something like that

File: openeo/Pre&PostCovid.ipynb [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "b4c667b8-a4f3-4dec-b28e-56b0e911a16e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# load the results\n",
+    "predata = xr.load_dataset(\"during_covid.nc\")\n",
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:256]: Note that you risk confusing yourself: you load *during* data as *predata*

File: openeo/Pre&PostCovid.ipynb [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    "ax2.plot(postdata.t, postdata.NO2.to_numpy().flatten(), color=\"r\")\n",
+    "ax2.set_xlabel(\"Post COVID\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5f459e68-b702-4dd8-a07b-79d5f285b6b9",
+   "metadata": {},
+   "source": [
+    "The red line in the post illustrates the NO2 concentration duration during COVID, and the green line represents the period after COVID for similar months of various years. Although the shift is not substantial, there's a minor decline in NO2 levels. \n",
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:320]:
```suggestion
    "The red line in the plot shows the NO2 concentration duration post COVID, and the green line represents the period during COVID for similar months of various years. Although the shift is not substantial, there's a minor decline in NO2 levels. \n",
```

I would also switch green and red subsentences

File: openeo/Pre&PostCovid.ipynb [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    " \n",
+    "Thus, as a reference to this notebook, additional scenarios can be explored using the Sentinel 5P data within the Copernicus Dataspace Ecosystem.\n",
+    "\n",
+    "The red line in the plot shows the NO2 concentration during the COVID period over the selected region in Delhi, while the green line represents the post-COVID period for equivalent months in various years. It shows that although the shift is not substantial, there's a minor decline in NO2 levels.\n",
+    "Thuslytaking this nonotebo as a reference caseok, further scenarios can be investigated using Sentinel 5P data within the Copernicus Data Space Ecosyst, such as PM2.5 concentration, ozone layer depletion, SO2 concentration, etcem."
```
Comment (soxofaan) [inline openeo/Pre&PostCovid.ipynb:325]: this last paragraph looks bad

File: openeo/Pre&PostCovid.ipynb (hunk 1 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: updated

File: openeo/Pre&PostCovid.ipynb (hunk 2 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: removed

File: openeo/Pre&PostCovid.ipynb (hunk 3 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: updated as suggested

File: openeo/Pre&PostCovid.ipynb (hunk 4 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: updated

File: openeo/Pre&PostCovid.ipynb (hunk 5 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: removed

File: openeo/Pre&PostCovid.ipynb (hunk 6 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: labelled as NO2

File: openeo/Pre&PostCovid.ipynb (hunk 7 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: maintained variable as duringdata

File: openeo/Pre&PostCovid.ipynb (hunk 8 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: I realised that earlier, I had made mistakes in presenting the result. Logically, it should decrease when no human activities going on.

Also switch the green and red representation as suggested(red-during, gree-post)

File: openeo/Pre&PostCovid.ipynb (hunk 9 above)
Comment (Pratichhya) [reply openeo/Pre&PostCovid.ipynb]: Ahh my bad, for some reason, it looked fine in markdown but not after executing the cell. It has been corrected

**Comment by Pratichhya** [comment]
> looks better now, thanks
> 
> fine for me to merge

Thanks to you too for your guidance 😄

**Pull request merged**